usr/src/uts/common/fs/zfs/sys/dnode.h

```diff
@@ -143,19 +150,37 @@ typedef struct dnode_phys {
 	uint8_t dn_flags;		/* DNODE_FLAG_* */
 	uint16_t dn_datablkszsec;	/* data block size in 512b sectors */
 	uint16_t dn_bonuslen;		/* length of dn_bonus */
-	uint8_t dn_pad2[4];
+	/*
+	 * dn_extra_slots is a placeholder for a feature in other ZFS
+	 * implementations. In this implementation, its value is always
+	 * 0. We declare it here to ensure it isn't used for a different
+	 * purpose, and to improve code portability with implementations
+	 * which support extra dnode slots.
+	 */
+	uint8_t dn_extra_slots;		/* # of subsequent slots consumed */
+	uint8_t dn_pad2[3];
 
 	/* accounting is protected by dn_dirty_mtx */
 	uint64_t dn_maxblkid;		/* largest allocated block ID */
 	uint64_t dn_used;		/* bytes (or sectors) of disk space */
 
+	/*
+	 * Both dn_pad2 and dn_pad3 are protected by the block's MAC. This
+	 * allows us to protect any fields that might be added here in the
+	 * future. In either case, developers will want to check
+	 * zio_crypt_init_uios_dnode() to ensure the new field is being
+	 * protected properly.
+	 */
 	uint64_t dn_pad3[4];
 
 	blkptr_t dn_blkptr[1];
 	uint8_t dn_bonus[DN_MAX_BONUSLEN - sizeof (blkptr_t)];
 	blkptr_t dn_spill;
 } dnode_phys_t;
 
+#define	DN_SPILL_BLKPTR(dnp)	(blkptr_t *)((char *)(dnp) + \
```

Comment:
Unfortunately, the GH PR mechanism won't let me highlight some of them (since they aren't changed as part of the PR), but with adding this, I've seen at least 1-2 instances where ZoL is using `DN_SPILL_BLKPTR` while OpenZFS is not (e.g. `dbuf_check_blkptr()`, `dbuf_write_ready()`) -- since it doesn't appear the extra slots are currently being used, I don't think it should currently cause an issue, though I wonder if it might cause future problems.